Applet/Applet/src/com/android/javacard/keymaster/KMKeymasterApplet.java

```diff
@@ -533,22 +620,124 @@ private void processAddRngEntropyCmd(APDU apdu) {
+    // read the cert chain from non-volatile memory. Cert chain is already in
+    // CBOR format.
+    seProvider.readCertificateChain(buffer, (short) (bufferStartOffset + 2));
+    // Encode cert chain.
+    encoder.encodeCertChain(buffer, bufferStartOffset, bufferLength);
+    sendOutgoing(apdu);
+  }
+
+  private void handleStateTransition() {
+    if (keymasterState == KMKeymasterApplet.FIRST_SELECT_STATE) {
+      if (provisionStatus == (PROVISION_STATUS_SIGN_KEY
+              | PROVISION_STATUS_SIGN_CERT_CHAIN | PROVISION_STATUS_SIGN_CERT_PARAMS)) {
+        keymasterState = KMKeymasterApplet.PROVISIONED_AKEY_ACERT_STATE;
+      }
+    } else if (keymasterState == KMKeymasterApplet.PROVISIONED_AKEY_ACERT_STATE) {
+      if (provisionStatus == (PROVISION_STATUS_SIGN_KEY
+              | PROVISION_STATUS_SIGN_CERT_CHAIN
+              | PROVISION_STATUS_SIGN_CERT_PARAMS | PROVISION_STATUS_ATTEST_IDS
+              | PROVISION_STATUS_SHARED_SECRET | PROVISION_STATUS_APPLET_PROVISIONED)) {
+        keymasterState = KMKeymasterApplet.PROVISIONED_AKEY_ACERT_AID_SS_STATE;
+      }
+    } else if (keymasterState == KMKeymasterApplet.PROVISIONED_AKEY_ACERT_AID_SS_STATE) {
+      keymasterState = KMKeymasterApplet.ACTIVE_STATE;
+    }
+  }
+
+  private void processProvisionAttestationCertParams(APDU apdu) {
+    receiveIncoming(apdu);
+    // Arguments
+    short blob = KMByteBlob.exp();
+    short argsProto = KMArray.instance((short) 3);
+    KMArray.cast(argsProto).add((short) 0, blob); // Cert - DER encoded issuer
+    KMArray.cast(argsProto).add((short) 1, blob); // Cert - Expiry Time
+    KMArray.cast(argsProto).add((short) 2, blob); // Cert - Auth Key Id
+    // Decode the argument.
+    short args = decoder.decode(argsProto, buffer, bufferStartOffset, bufferLength);
+
+    // save issuer - DER Encoded
+    tmpVariables[0] = KMArray.cast(args).get((short) 0);
+    repository.setIssuer(
+        KMByteBlob.cast(tmpVariables[0]).getBuffer(),
+        KMByteBlob.cast(tmpVariables[0]).getStartOff(),
+        KMByteBlob.cast(tmpVariables[0]).length());
+
+    // save expiry time - UTC or General Time - YYMMDDhhmmssZ or YYYYMMDDhhmmssZ.
+    tmpVariables[0] = KMArray.cast(args).get((short) 1);
+    repository.setCertExpiryTime(
+        KMByteBlob.cast(tmpVariables[0]).getBuffer(),
+        KMByteBlob.cast(tmpVariables[0]).getStartOff(),
+        KMByteBlob.cast(tmpVariables[0]).length());
+
+    // Auth Key Id - from cert associated with imported attestation key.
+    tmpVariables[0] = KMArray.cast(args).get((short) 2);
+    repository.setAuthKeyId(
+        KMByteBlob.cast(tmpVariables[0]).getBuffer(),
+        KMByteBlob.cast(tmpVariables[0]).getStartOff(),
+        KMByteBlob.cast(tmpVariables[0]).length());
+
+    provisionStatus |= KMKeymasterApplet.PROVISION_STATUS_SIGN_CERT_PARAMS;
+    handleStateTransition();
+    sendError(apdu, KMError.OK);
+  }
+
+  private void processProvisionAttestationCertChainCmd(APDU apdu) {
```

Comment:
HAL should send BSTR(CERT_CHAIN) instead of ARRAY(CERT_CHAIN)
ONly validate the length of BSTR. 
Make a configurable class.